samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaConsumerProxy.java

```diff
@@ -0,0 +1,444 @@
+
+  private volatile boolean isRunning = false;
+  private volatile Throwable failureCause = null;
+  private final CountDownLatch consumerPollThreadStartLatch = new CountDownLatch(1);
+
+  // package private constructor
+  KafkaConsumerProxy(Consumer<K, V> kafkaConsumer, String systemName, String clientId,
+      KafkaSystemConsumer.KafkaConsumerMessageSink messageSink, KafkaSystemConsumerMetrics samzaConsumerMetrics,
+      String metricName) {
+
+    this.kafkaConsumer = kafkaConsumer;
+    this.systemName = systemName;
+    this.sink = messageSink;
+    this.kafkaConsumerMetrics = samzaConsumerMetrics;
+    this.metricName = metricName;
+    this.clientId = clientId;
+
+    this.kafkaConsumerMetrics.registerClientProxy(metricName);
+
+    consumerPollThread = new Thread(createProxyThreadRunnable());
+    consumerPollThread.setDaemon(true);
+    consumerPollThread.setName(
+        "Samza KafkaConsumerProxy Poll " + consumerPollThread.getName() + " - " + systemName);
+  }
+
+  @Override
+  public String toString() {
+    return String.format("consumerProxy-%s-%s", systemName, clientId);
+  }
+
+  public void start() {
+    if (!consumerPollThread.isAlive()) {
+      LOG.info("Starting KafkaConsumerProxy polling thread for system " + systemName + " " + this.toString());
+
+      consumerPollThread.start();
+
+      // we need to wait until the thread starts
+      while (!isRunning && failureCause == null) {
+        try {
+          consumerPollThreadStartLatch.await(3000, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+          LOG.info("Got InterruptedException", e);
+        }
+      }
+    } else {
+      LOG.warn("Tried to start an already started KafkaConsumerProxy (%s). Ignoring.", this.toString());
+    }
+
+    if (topicPartitions2SSP.size() == 0) {
+      String msg = String.format("Cannot start empty set of TopicPartitions for system %s, clientid %s",
+          systemName, clientId);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+  }
+
+  /**
+   * Stop the thread and wait for it to stop.
+   * @param timeoutMs how long to wait in join
+   */
+  public void stop(long timeoutMs) {
+    LOG.info("Shutting down KafkaConsumerProxy poll thread:" + consumerPollThread.getName());
+
+    isRunning = false;
+    try {
+      consumerPollThread.join(timeoutMs);
+      // join returns event if the thread didn't finish
+      // in this case we should interrupt it and wait again
+      if (consumerPollThread.isAlive()) {
+        consumerPollThread.interrupt();
+        consumerPollThread.join(timeoutMs);
+      }
+    } catch (InterruptedException e) {
+      LOG.warn("Join in KafkaConsumerProxy has failed", e);
+      consumerPollThread.interrupt();
+    }
+  }
+
+  /**
+   * Add new partition to the list of polled partitions.
+   * This method should be called only at the beginning, before the thread is started.
+   */
+  public void addTopicPartition(SystemStreamPartition ssp, long nextOffset) {
```

Comment:
I think these methods can all be package private (I mean remove `public`, not add the /* package private */ annotation) since the only user should be KafkaSystemConsumer